File: Chapter04/Catapult.c
const tSensors touchSensor1 = (tSensors) S1;
const tSensors touchSensor2 = (tSensors) S2;
task main() {
	// endless loop
	while(true) {
		// lift crank mechanism
		nMotorEncoder[motorA] = 0;
		while(nMotorEncoder[motorA] <= 30 ) {
			motor[motorA] = 75;
		}
		motor[motorA] = 0;
		// load catapult
		// run motor until until the first
		// touch sensor gets pressed
		while(SensorValue(touchSensor1) == 0) {
			motor[motorC] = 100;
		};
		motor[motorC] = 0;
		// wait for the user to press the second touch sensor
		while(SensorValue(touchSensor2) == 0) {
			// do nothing
		};
		// drop crank mechanism
		nMotorEncoder[motorA] = 0;
		while(nMotorEncoder[motorA] >= -30 ) {
			motor[motorA] = -75;
		}
		motor[motorA] = 0;
		// wait for two seconds
		wait10Msec(200);
	};
}

File: Chapter04/Catapult.lua
-- function Catapult
function Catapult()
	-- configure the touch sensors on input ports 1 and 2
	nxt.InputSetDigi0(1)
	nxt.InputSetDirOutDigi0(1)
	nxt.InputSetDigi0(2)
	nxt.InputSetDirOutDigi0(2)
	-- configure motors A and C to run in "brake" mode
	nxt.OutputSetMode(1,2)
	nxt.OutputSetMode(3,2)
	-- loop forever
	while 1 do
		-- lift crank mechanism
		-- rotate motor A by 30 degrees
		rotate(1,30)
		-- load catapult
		-- run motor C until until the first
		-- touch sensor gets pressed by the liftarm
		nxt.outputSetSpeed(3, 32, 100, 0)
		while nxt.InputGetRawAd(1) == 0
			-- do nothing
		end
		nxt.OutputSetSpeed(3,0,0,0)
		-- wait for the user to press the second touch sensor
		while nxt.InputGetRawAd(2) == 0
			-- do nothing
		end
		-- drop crank mechanism
		rotate(1,-30)
		-- wait for 2 seconds
		local endOfWait = nxt.TimerRead() + 2000
		while nxt.TimerRead() < endOfWait
			-- do nothing
		end
	end
end

-- function to rotate a motor by a given number of degrees
function rotate(motor,degrees)
	nxt.OutputEnableRegulation(motor,1)
	nxt.OutputSetMode(motor,2)
	nxt.OutputSetSpeed(motor,32,75,0)
	nxt.OutputSetTachoLimit(motor,degrees)
end

-- now run the Catapult function
Catapult()
